src/main/java/emu/grasscutter/game/gacha/GachaManager.java

```diff
@@ -183,11 +184,28 @@ private synchronized int drawRoulette(int[] weights, int cutoff) {
 
 	private synchronized int doRarePull(int[] featured, int[] fallback1, int[] fallback2, int rarity, GachaBanner banner, PlayerGachaBannerInfo gachaInfo) {
 		int itemId = 0;
-		boolean pullFeatured = (gachaInfo.getFailedFeaturedItemPulls(rarity) >= 1)  // Lost previous coinflip
+		boolean pullFeatured = (gachaInfo.getFailedFeaturedItemPulls(rarity) >= (banner.hasEpitomized() ? banner.getWishMaxProgress() : 1))  // Lost previous coinflip
```

Comment:
1) This appears to be missing a `rarity` check
2) I think this breaks the guarantee of getting a featured item after losing a coinflip/50:50 - need another counter instead of reusing `FailedFeaturedItemPulls`. Maybe it could be called `FailedChosenItemPulls`?

It's probably cleaner to move the check for `WishMaxProgress` to the start, and check if the item matches at the end, something like this:
```
...{
int itemId = 0;
boolean epitomized = (banner.hasEpitomized() && (rarity == 5));
if (epitomized && (gachaInfo.getFailedChosenItemPulls() >= banner.getWishMaxProgress())) {
    itemId = gachaInfo.getWishItemId(); // Pick the chosen item
} else {
    boolean pullFeatured = ...
    if (pullFeatured && (featured.length > 0)) {
        // Pick a featured item
    } else {
        // Pick a fallback item
    }
}
if (epitomized && (itemId != gachaInfo.getWishItemId())
    gachaInfo.setFailedChosenItemPulls(0);
else
    gachaInfo.addFailedChosenItemPulls(1);

return itemId;
}
```

Reply:
okay thanks, i will update soon